File: src/main/java/com/example/parking/domain/reservation/entity/Reservation.java
package com.example.parking.domain.reservation.entity;

import com.example.parking.domain.parkingLot.entity.ParkingLot;
import com.example.parking.domain.parkingspot.entity.ParkingSpot;
import com.example.parking.domain.user.entity.User;
import jakarta.persistence.*;
import lombok.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDateTime;

@Entity
@Table(name = "reservations")
@Getter
@Builder
@AllArgsConstructor //빌더가 내부적으로 사용하는 '모든 필드 생성자' 생성
@NoArgsConstructor(access = AccessLevel.PROTECTED) //JPA가 사용하는 '기본 생성자' 생성
@EntityListeners(AuditingEntityListener.class)
public class Reservation {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "reservation_id")
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parking_lot_id", nullable = false)
    private ParkingLot parkingLot;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parking_spot_id", nullable = false)
    private ParkingSpot parkingSpot;

    @Column(name = "parking_start_time", nullable = false)
    private LocalDateTime startTime;

    @Column(name = "parking_end_time", nullable = false)
    private LocalDateTime endTime;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false)
    private ReservationStatus status;

    @CreatedDate
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "canceled_at")
    private LocalDateTime canceledAt;

    // Builder에 DB 필수값(nullable = false)인 주차장과 자리 파라미터 추가 완료
    @Builder
    public Reservation(User user, ParkingLot parkingLot, ParkingSpot parkingSpot, LocalDateTime startTime, LocalDateTime endTime) {
        this.user = user;
        this.parkingLot = parkingLot;
        this.parkingSpot = parkingSpot;
        this.startTime = startTime;
        this.endTime = endTime;
        this.status = ReservationStatus.PENDING;
    }

    private LocalDateTime paymentRequestedAt; // 결제 버튼 클릭 시점

    //결제 호출하기
    public void startPayment() {
        this.paymentRequestedAt = LocalDateTime.now();
    }

    //예약 상태를 cancel로 변경
    public void cancel() {
        this.status = ReservationStatus.CANCELED;
        this.canceledAt = LocalDateTime.now();
    }

    //예약 상태를 CONFIRMED로 변경
    public void confirm() {
        this.status = ReservationStatus.CONFIRMED;
    }

    //예약 상태를 COMPLETED로 변경
    public void complete() {
        this.status = ReservationStatus.COMPLETED;
    }

    public void pending() {
        this.status = ReservationStatus.PENDING;
    }

}

File: src/main/java/com/example/parking/domain/reservation/entity/Reservation.kt
package com.example.parking.domain.reservation.entity

import com.example.parking.domain.parkingLot.entity.ParkingLot
import com.example.parking.domain.parkingspot.entity.ParkingSpot
import com.example.parking.domain.user.entity.User
import jakarta.persistence.*
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import java.time.LocalDateTime

@Entity
@Table(name = "reservations")
@EntityListeners(AuditingEntityListener::class)
class Reservation protected constructor(

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    val user: User,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parking_lot_id", nullable = false)
    val parkingLot: ParkingLot,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parking_spot_id", nullable = false)
    val parkingSpot: ParkingSpot,

    @Column(name = "parking_start_time", nullable = false)
    val startTime: LocalDateTime,

    @Column(name = "parking_end_time", nullable = false)
    val endTime: LocalDateTime,

    ) {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "reservation_id")
    var id: Long? = null
        protected set

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false)
    var status: ReservationStatus = ReservationStatus.PENDING
        protected set

    @CreatedDate
    @Column(name = "created_at", nullable = false, updatable = false)
    var createdAt: LocalDateTime? = null
        protected set

    @Column(name = "canceled_at")
    var canceledAt: LocalDateTime? = null
        protected set

    @Column(name = "payment_requested_at")
    var paymentRequestedAt: LocalDateTime? = null
        protected set

    fun startPayment() {
        this.paymentRequestedAt = LocalDateTime.now()
    }

    fun cancel() {
        this.status = ReservationStatus.CANCELED
        this.canceledAt = LocalDateTime.now()
    }

    fun confirm() {
        this.status = ReservationStatus.CONFIRMED
    }

    fun complete() {
        this.status = ReservationStatus.COMPLETED
    }

    fun finish() {
        this.status = ReservationStatus.FINISHED
    }

    fun pending() {
        this.status = ReservationStatus.PENDING
    }

    companion object {
        fun of(
            user: User,
            parkingLot: ParkingLot,
            parkingSpot: ParkingSpot,
            startTime: LocalDateTime,
            endTime: LocalDateTime
        ) = Reservation(
            user = user,
            parkingLot = parkingLot,
            parkingSpot = parkingSpot,
            startTime = startTime,
            endTime = endTime
        )
    }
}